core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,226 @@
+  static {
+    ROOT.close();
+  }
+
+  /**
+   * A listener notified on context {@link #close()}.
+   */
+  public interface Listener {
+    /**
+     * @param context the newly closed context.
+     */
+    public void closed(Context context);
+  }
+
+  /**
+   * Create a new context with the given key value set.
+   */
+  public <V> Context withValue(Key<V> k1, V v1) {
+    throw new UnsupportedOperationException();
+  }
+
+  public <V1, V2> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2) {
+    throw new UnsupportedOperationException();
+  }
+
+  public <V1, V2, V3> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2, Key<V3> k3, V3 v3) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Create a new context with the given listener
+   */
+  public <V> Context withListener(Listener listener) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Create a new context that does not cascade closure from its parent
+   */
+  public <V> Context withNoCascade() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Create a new context that cascades closure from its parent
+   */
+  public <V> Context withCascade() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Return the context associated with the current thread, will never return {@code null} as
+   * the {@link #ROOT} context is implicitly associated with all threads.
+   */
+  public static Context current() {
+    return null;
+  }
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context. It is perfectly fine to attach a
+   * context which is already closed.
+   *
+   * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+   * @return the {@link #current()} context after {@code this} has been detached.
+   */
+  // Discussion point: is the return useful? inconsistency with other fluent methods if we have them
+  public Context detach() {
+    throw new UnsupportedOperationException();
+  }
+
+
+  /**
+   * Attach this context to the thread and make it current, the previously current context
+   * will be restored when detach is called.
+   * @return
+   */
+  // Discussion point, make fluent?
+  public void attach() {
+    throw new UnsupportedOperationException();
+  }
+
+
+  /**
+   * Close this context and optionally provide a cause for the closure, all listeners are notified
+   * immediately on the calling thread
+   *
+   * @return {@code true} if this call closed the context and notified listeners, {@code false} if
+   *          the context was already closed.
+   */
+  public boolean close(@Nullable Throwable cause) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Close the context normally.
+   */
+  @Override
+  // Discussion point: Cant return boolean here and be Closeable
+  public void close() {
```

Comment:
Added attachAsCloseable() and detachAndClose() the former for

try (AutoCloseable c = Context.current().withValue().attachAsCloseable()) {
    ....
}

and the latter for more old school

Context mine = Context.current().withValue();
Throwable cause = null;
try {
   mine.attach();
   <do work>
} catch (Throwable t) {
  cause = t;
} finally {
  c.detachAndClose(cause);
}

which is a little ugly, I could be persuaded that

Context mine = Context.current().withValue();
try {
   mine.attach();
   <do work>
} catch (Throwable t) {
  mine.close(t);
} finally {
  //Safe to close twice
  c.detachAndClose(); 
}

is preferable